test/test_testing.py

```diff
@@ -4738,6 +4738,94 @@ def test_storage_mount(self):
         self.harness.attach_storage(storage_id)
         self.assertTrue((self.root / "mounts/foo/bar").read_text(), "foobar")
 
+    def test_storage_attach(self):
+        class MyCharm(ops.CharmBase):
+            def __init__(self, framework: ops.Framework):
+                super().__init__(framework)
+                self.attached: typing.List[str] = []
+                self.locations: typing.List[pathlib.Path] = []
+                framework.observe(self.on['test-storage'].storage_attached, self._on_attach)
+
+            def _on_attach(self, event: ops.StorageAttachedEvent):
+                self.attached.append(event.storage.full_id)
+                self.locations.append(event.storage.location)
+
+        meta = '''
+            name: test
+            containers:
+                test-container:
+                    mounts:
+                        - storage: test-storage
+                          location: /mounts/foo
+            storage:
+                test-storage:
+                    type: filesystem
+            '''
+        # If `begin()` hasn't been called, `attach` does not emit storage-attached.
+        harness = ops.testing.Harness(MyCharm, meta=meta)
+        self.addCleanup(harness.cleanup)
+        harness.add_storage('test-storage', attach=True)
+        harness.begin()
+        self.assertNotIn('test-storage/0', harness.charm.attached)
+
+        # `attach` doesn't emit storage-attached before `begin_with_initial_hooks`.
+        harness = ops.testing.Harness(MyCharm, meta=meta)
+        self.addCleanup(harness.cleanup)
```

Comment:
It seems a bit weird to adding cleanups for multiple harness instances, which will all happen at the end. I don't think it's actually a problem (each has a random temp directory), but still feel a bit dirty. I think we should either just call `harness.cleanup()` manually after each, or probably better, have one test per harness / thing we're checking (but re-use the charm and metadata).

Reply:
Manually after each really ought to be done in finally blocks since it's possible the test fails and we don't want to leave things around in that case. Splitting into separate tests is fine; I'll do that.